File: Simple Linear Regression/Delivery.py
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy import stats
import pylab


# Opening file and assigning to an object
delivery = pd.read_csv('C:/Users/lnram/Desktop/ExcelR/Module 6 - Simple Linear Regression/Assignments/delivery_time.csv')

# Business Problem: Predict delivery_time with Max accuracy

#EDA/Descriptive Analytics:
# First Moment Business Decision - Measure of Central Tendancy

delivery.describe()

# Second Moment Business Decision - Measure of Dispersion

delivery.columns = "delivery_time" , "sort_time"

#Std Deviation
print("Std Deviation of delivery_time: " , delivery.delivery_time.std())
print("Std Deviation of sort_time: " , delivery.sort_time.std())


#Range
print("Range of delivery_time: " , max(delivery.delivery_time) - min(delivery.delivery_time))
print("Range of sort_time: " , max(delivery.sort_time) - min(delivery.sort_time))


# Third Moment Business Decision - Skewness
print("Skewness of delivery_time : " , delivery.delivery_time.skew())
print("Skewness of sort_time : " , delivery.sort_time.skew())


# Fourth Moment Business Decision - Kurtosis
print("Kurtosis of delivery_time : " , delivery.delivery_time.kurt())
print("Kurtosis of sort_time : " , delivery.sort_time.kurt())

# Graphical Representation - Distribution, Box Plot, Scatter Plot
sns.distplot(delivery.delivery_time)
sns.distplot(delivery.sort_time)

plt.boxplot(delivery.delivery_time)
plt.boxplot(delivery.sort_time)

plt.scatter(delivery.sort_time,delivery.delivery_time)

#Modeling - Linear Regression

import statsmodels.formula.api as smf

model1 = smf.ols('delivery_time ~ sort_time' , data = delivery).fit()
model1.summary()  # output Evaluation

# Model Evaluation & Assumptions

np.corrcoef(delivery.delivery_time,delivery.sort_time) # Correlation Coefficient

error = model1.predict(delivery.sort_time) - delivery.delivery_time
rmse = np.sqrt(np.mean(error * error)) # Root Mean Squared Value
rmse


stats.shapiro(error) # Shapiro test for normality
stats.probplot(error , dist = "norm" , plot = pylab) # QQplot

# Coeffs p-value is less than 0.05 - Coeffs values are stastically significant
# F Statistic p-value < 0.05 - Regression Eqn is statistically significant
# RMSE is very low
# Shapiro test on Residuals - p-val > 0.05, Residuals are normal
# But the R Squared value is only 0.68


#Modeling - Linear Regression - Log


np.corrcoef(delivery.delivery_time,np.log(delivery.sort_time)) # Correlation Coefficient
plt.scatter(delivery.delivery_time,np.log(delivery.sort_time))


model2 = smf.ols('delivery_time ~ np.log(sort_time)' , data = delivery).fit()
model2.summary()  # output Evaluation

# Model Evaluation & Assumptions
error2 = model2.predict(delivery.sort_time) - delivery.delivery_time
rmse2 = np.sqrt(np.mean(error2 * error2)) # Root Mean Squared Value
rmse2


stats.shapiro(error2) # Shapiro test for normality
stats.probplot(error2 , dist = "norm" , plot = pylab) # QQplot

# Coeffs p-value is less than 0.05 - Coeffs values are stastically significant
# F Statistic p-value < 0.05 - Regression Eqn is statistically significant
# RMSE is very low
# Shapiro test on Residuals - p-val > 0.05, Residuals are normal
# But the R Squared value is only 0.69



#Modeling - Linear Regression - Exponential


np.corrcoef(np.log(delivery.delivery_time),delivery.sort_time) # Correlation Coefficient
plt.scatter(np.log(delivery.delivery_time), delivery.sort_time)

model3 = smf.ols('np.log(delivery_time) ~ sort_time' , data = delivery).fit()
model3.summary()  # output Evaluation

# Model Evaluation & Assumptions
error3 = np.exp(model3.predict(delivery.sort_time)) - delivery.delivery_time
rmse3 = np.sqrt(np.mean(error3 * error3)) # Root Mean Squared Value
rmse3


stats.shapiro(error3) # Shapiro test for normality
stats.probplot(error3 , dist = "norm" , plot = pylab) # QQplot


# Coeffs p-value is less than 0.05 - Coeffs values are stastically significant
# F Statistic p-value < 0.05 - Regression Eqn is statistically significant
# RMSE is very low
# Shapiro test on Residuals - p-val > 0.05, Residuals are normal
# R Squared value is 0.71



#Modeling - Polynomial Regression


np.corrcoef(delivery.delivery_time,delivery.sort_time) # Correlation Coefficient
plt.scatter(delivery.delivery_time, delivery.sort_time)

model4 = smf.ols('delivery_time ~ sort_time + I(sort_time * sort_time)' , data = delivery).fit()
model4.summary()  # output Evaluation

# Model Evaluation & Assumptions
error4 = model4.predict(delivery.sort_time) - delivery.delivery_time
rmse4 = np.sqrt(np.mean(error4 * error4)) # Root Mean Squared Value
rmse4


stats.shapiro(error4) # Shapiro test for normality
stats.probplot(error4 , dist = "norm" , plot = pylab) # QQplot




# Coeffs p-value > 0.05 - Coeffs values are stastically insignificant
# F Statistic p-value < 0.05 - Regression Eqn is statistically significant
# RMSE is very low
# Shapiro test on Residuals - p-val > 0.05, Residuals are normal
# R Squared value is 0.69



model5 = smf.ols('np.log(delivery_time) ~ np.log(sort_time)' , data = delivery).fit()
model5.summary()  # output Evaluation

# Model Evaluation & Assumptions
error5 = model5.predict(delivery.sort_time) - delivery.delivery_time
rmse5 = np.sqrt(np.mean(error5 * error5)) # Root Mean Squared Value
rmse5


stats.shapiro(error5) # Shapiro test for normality
stats.probplot(error5 , dist = "norm" , plot = pylab) # QQplot

# Coeffs p-value is less than 0.05 - Coeffs values are stastically significant
# F Statistic p-value < 0.05 - Regression Eqn is statistically significant
# RMSE is very low
# Shapiro test on Residuals - p-val > 0.05, Residuals are normal
# R Squared value is 0.77

final_model = smf.ols('np.log(delivery_time) ~ np.log(sort_time)' , data = delivery).fit()
final_model.summary()  # output Evaluation


final_model.conf_int(0.05) # CI of Coeffs at 95% limit
final_model.predict(pd.DataFrame(delivery.sort_time)) # Predicting delivery_time based on confidence intervals

File: Simple Linear Regression/Delivery.R
# Opening file and assigning to an object
delivery <- read.csv(choose.files())
View(delivery)

# Business Problem: Predict delivery_time with Max accuracy

#EDA/Descriptive Analytics:
# First Moment Business Decision - Measure of Central Tendancy

summary(delivery)

# Second Moment Business Decision - Measure of Dispersion

colnames(delivery) <- c("delivery_time" , "sort_time")
attach(delivery)

#Std Deviation
print(paste("Std Deviation of delivery_time: " , sd(delivery_time)))
print(paste("Std Deviation of sort_time: " , sd(sort_time)))

#Variance
print(paste("Variance of delivery_time: " , sd(delivery_time)^2))
print(paste("Variance of sort_time: " , sd(sort_time)^2))

#Range
print(paste("Range of delivery_time: " , max(delivery_time) - min(delivery_time)))
print(paste("Range of sort_time: " , max(sort_time) - min(sort_time)))

# Third Moment Business Decision - Skewness
library(moments)
print(paste("Skewness of delivery_time : " , skewness(delivery_time)))
print(paste("Skewness of sort_time : " , skewness(sort_time)))

# Fourth Moment Business Decision - Kurtosis
print(paste("Kurtosis of delivery_time : " , kurtosis(delivery_time)))
print(paste("Kurtosis of sort_time : " , kurtosis(sort_time)))

# Graphical Representation - Distribution, Box Plot, Scatter Plot
plot(density(delivery_time))
plot(density(sort_time))

boxplot(delivery_time)
boxplot(sort_time)

plot(sort_time,delivery_time)

#Modeling - Linear Regression

model1 <- lm(delivery_time ~ sort_time)
summary(model1) # output Evaluation

# Model Evaluation & Assumptions

cor(delivery_time,sort_time) # Correlation Coefficient

error <- predict(model1) - delivery_time
rmse <- sqrt(mean(error^2)) # Root Mean Squared Value
rmse

shapiro.test(model1$residuals) # Shapiro test for normality
qqnorm(model1$residuals) # QQplot
qqline(model1$residuals) # QQLine

# Coeffs p-value is less than 0.05 - Coeffs values are stastically significant
# F Statistic p-value < 0.05 - Regression Eqn is statistically significant
# RMSE is very low
# Shapiro test on Residuals - p-val > 0.05, Residuals are normal
# But the R Squared value is only 0.68


#Modeling - Linear Regression - Log


cor(delivery_time,log(sort_time)) # Correlation Coefficient
plot(delivery_time,log(sort_time))

model2 <- lm(delivery_time ~ log(sort_time))
summary(model2) # output Evaluation

# Model Evaluation & Assumptions

error <- predict(model2) - delivery_time
rmse <- sqrt(mean(error^2)) # Root Mean Squared Value
rmse

shapiro.test(model2$residuals) # Shapiro test for normality
qqnorm(model2$residuals) # QQplot
qqline(model2$residuals) # QQLine

# Coeffs p-value is less than 0.05 - Coeffs values are stastically significant
# F Statistic p-value < 0.05 - Regression Eqn is statistically significant
# RMSE is very low
# Shapiro test on Residuals - p-val > 0.05, Residuals are normal
# But the R Squared value is only 0.69



#Modeling - Linear Regression - Exponential


cor(log(delivery_time),sort_time) # Correlation Coefficient
plot(log(delivery_time), sort_time)

model3 <- lm(log(delivery_time) ~ sort_time)
summary(model3) # output Evaluation

# Model Evaluation & Assumptions

error <- exp(predict(model3 , level = "confidence")) - delivery_time
rmse <- sqrt(mean(error^2)) # Root Mean Squared Value
rmse

shapiro.test(model3$residuals) # Shapiro test for normality
qqnorm(model3$residuals) # QQplot
qqline(model3$residuals) # QQLine

# Coeffs p-value is less than 0.05 - Coeffs values are stastically significant
# F Statistic p-value < 0.05 - Regression Eqn is statistically significant
# RMSE is very low
# Shapiro test on Residuals - p-val > 0.05, Residuals are normal
# R Squared value is 0.71



#Modeling - Polynomial Regression


cor(delivery_time,sort_time) # Correlation Coefficient
plot(delivery_time, sort_time)

model4 <- lm(delivery_time ~ sort_time + I(sort_time * sort_time))
summary(model4) # output Evaluation

# Model Evaluation & Assumptions

error <- predict(model4 , level = "confidence") - delivery_time
rmse <- sqrt(mean(error^2)) # Root Mean Squared Value
rmse

shapiro.test(model4$residuals) # Shapiro test for normality
qqnorm(model4$residuals) # QQplot
qqline(model4$residuals) # QQLine

# Coeffs p-value > 0.05 - Coeffs values are stastically insignificant
# F Statistic p-value < 0.05 - Regression Eqn is statistically significant
# RMSE is very low
# Shapiro test on Residuals - p-val > 0.05, Residuals are normal
# R Squared value is 0.69




#Modeling - Linear Regression - Log Transform of X & Y 


cor(log(delivery_time),log(sort_time)) # Correlation Coefficient
plot(log(delivery_time), log(sort_time))

model5 <- lm(log(delivery_time) ~ log(sort_time))
summary(model5) # output Evaluation

# Model Evaluation & Assumptions

error <- exp(predict(model3 , level = "confidence")) - delivery_time
rmse <- sqrt(mean(error^2)) # Root Mean Squared Value
rmse

shapiro.test(model3$residuals) # Shapiro test for normality
qqnorm(model3$residuals) # QQplot
qqline(model3$residuals) # QQLine

# Coeffs p-value is less than 0.05 - Coeffs values are stastically significant
# F Statistic p-value < 0.05 - Regression Eqn is statistically significant
# RMSE is very low
# Shapiro test on Residuals - p-val > 0.05, Residuals are normal
# R Squared value is 0.77

final_model <- lm(log(delivery_time) ~ log(sort_time))
summary(final_model)

confint(final_model) # CI of Coeffs at 95% limit
predict(final_model , interval = "confidence") # Predicting delivery_time based on confidence intervals
